drivers/flash/spi_nor.c

```diff
@@ -1246,6 +1254,44 @@ static int spi_nor_configure(const struct device *dev)
 	return 0;
 }
 
+#ifdef CONFIG_PM_DEVICE
+
+static int spi_nor_pm_control(const struct device *dev, enum pm_device_action action)
+{
+	int rc = 0;
+
+	switch (action) {
+#ifdef CONFIG_SPI_NOR_IDLE_IN_DPD
+	case PM_DEVICE_ACTION_SUSPEND:
+	case PM_DEVICE_ACTION_RESUME:
+		break;
+#else
+	case PM_DEVICE_ACTION_SUSPEND:
+		acquire_device(dev);
+		rc = enter_dpd(dev);
+		release_device(dev);
```

Comment:
In case when `CONFIG_SPI_NOR_IDLE_IN_DPD` is enabled, this wakes up device in `acquire_device` then puts it to sleep twice.

Reply:
If that is enabled this code isn't run, see the enclosing #ifdef. Same for the second location

Reply:
Ah sorry, somehow missed that.
Shouldn't this check `k_sem_count_get` before putting device to sleep? Because there may be something scheduled just after PM event, and it will wake up device anyway.
In case this locks on `acquire_device`, is it possible that more than one PM event get scheduled?

Reply:
Checking the semaphore isn't required because it is only taken for the duration of an API call (if it is held longer that is a bug), and therefore our own call to `acquire_device` ensures that no other action is running.

The higher level "Device PM" API is responsible for ensuring that PM events are only run in the correct state, and "Device Runtime PM" is the correct API to use if you want multiple users for the driver with PM.